Refactor+Bugfix/SW 109 Race condition bug during multi step rollbacks

amrsoll opened the pull request:
See commit messages

amrsoll committed:
Refactor/Split responsabilities and add tests

- Split the cli parsing and handling (at least partially)
- Set the default color set in an Enum
- sort imports
- state history is a self-limiting buffer
- Add a `stop()` function to gracefully end the handling loop
- Remove unnecessary comments
- include pytest in the extra "develop" requirements

amrsoll committed:
Refactor/Separate settings from other commands

- Parse settings from the same function as the Command parser
- Raise an Exception  that wasn't handled in the socket monitoring loop

amrsoll committed:
Refactor/Sleep for a frame in the handle function.

amrsoll committed:
Refactor/Various

- bgr -> rgb to reflect the correct order of the colors in array
- parse8bit() change logic to truncate instead of putting a ceiling on the value
- parse8bit() logs an error in case a string value cannot be parsed
- _mylinspace() changed to an iterator instead of returning a list
- fade_off() uses fewer magic numbers.

amrsoll committed:
Fix+Refactor/Add COMMANDS_REQ_ARG

- Forgot to add the list of commnads that require at least an argument
- Refactor the list of LED IDs with a readable range()

amrsoll committed:
Remove unused focus tool colors

amrsoll committed:
Remove all focus tool logic (unused)

amrsoll committed:
Fix rollback bug.

- Adds tests fr the rollback method.

amrsoll committed:
Fix rollback bug

Add tests to verify rollback stability

ahmed-mrbeam commented on mrbeam_ledstrips/server.py:
```diff
@@ -187,17 +187,25 @@ def start(self):
+		self._socket_monitor(self.server_address, callback=self.parse_socket_msg)
 
-	def on_state_change(self, state):
+	def parse_socket_msg(self, command):
```
Please add doc string

ahmed-mrbeam commented on mrbeam_ledstrips/tests/test_manually.py:
```diff
@@ -0,0 +1,89 @@
+"""Duration for which to test the animations"""
+
+tests_dir_path = path.dirname(path.realpath(__file__))
+PNG_DIR = path.join(tests_dir_path, "../..", "extras", "png")
```
I would suggest to always use os.path instead of path. for more clarity

ahmed-mrbeam commented on mrbeam_ledstrips/tests/test_manually.py:
```diff
@@ -0,0 +1,89 @@
+
+from . import LEDLoopTester
+
+ANIMATION_TIME = .5
```
good practice is to mention in the constant name the unit of time used, ms, ns, sec, min, ....
I suggest : `ANIMATION_TIME_IN_SEC`

ahmed-mrbeam commented on mrbeam_ledstrips/tests/test_manually.py:
```diff
@@ -0,0 +1,89 @@
+tests_dir_path = path.dirname(path.realpath(__file__))
+PNG_DIR = path.join(tests_dir_path, "../..", "extras", "png")
+
+def user_accept(prompt):
```
doc string

ahmed-mrbeam commented on mrbeam_ledstrips/server.py:
```diff
@@ -187,17 +187,25 @@ def start(self):
+			available_settings = list(chain(SETTINGS.values()))
+			_split = command.split(":")
+			if len(_split) <= 2:
+				return "ERROR : Need to specifiy set:SETTING:VAL where SETTING is \{{}\}".format(",".join(available_settings))
```
why returning a string instead of raising an error or returning an error code ?

amrsoll replied:
In this case I was preserving previous logic

amrsoll replied:
See state_animations.py

ahmed-mrbeam commented on mrbeam_ledstrips/server.py:
```diff
@@ -187,17 +187,25 @@ def start(self):
 		else:
-			response = self.leds.change_state(state)
-		return response
+			return self.leds.change_state(command)
```
so, what exactly the command we expect in this case? don't we need a sanity check here as well ?

Usually else is for unsupported commands

amrsoll replied:
the change_state() function does that check. Anything else than strings starting with `set:` or asking for info (`?` / `info`) is used as an argument for changing the states

amrsoll replied:
would you prefer a regex verification of the input?

amrsoll replied:
more specifically:
I moved the `set_settings()` logic out of the `change_state()` function.

ahmed-mrbeam commented on mrbeam_ledstrips/server.py:
```diff
@@ -172,10 +173,9 @@ def _socket_monitor(self, server_address, callback):
-			pass
-		except Exception:
-			self.logger.exception("Exception in socket monitor: ")
+		except Exception as e:
```
so now we are ok with raising KeyboardInterrupt, SystemExit ?

amrsoll replied:
I didn't mention it in the commits - Given that I could read or imagine a reason for why we want to catch these, I preferred to remove them because they only hinder us in trying to kill the LED server when we would like.

ahmed-mrbeam commented on mrbeam_ledstrips/server.py:
```diff
@@ -187,17 +187,25 @@ def start(self):
+		elif command.startswith("set:"):
+			available_settings = list(chain(SETTINGS.values()))
+			_split = command.split(":")
+			if len(_split) <= 2:
```
what is 2, magic number

amrsoll replied:
Not so magic considering the following message

amrsoll commented:
@ahmed-mrbeam did you forget to look at `mrbeam_ledstrips/state_animations.py` or did you have nothing to add to it?

amrsoll committed:
Refactor/Use os.path and declare physical units.

amrsoll committed:
Add docstrings

- rm commented code
- fewer code repeats.

amrsoll committed:
Remove 'enum' as a dependency

I didn't figure it was a built-in lib.

amrsoll committed:
Add docstrings.

amrsoll committed:
Refactor/Move basic functions out of LED class.

amrsoll committed:
Fix _my_linspace.

amrsoll commented:
I highly encourage testing it using 

On the 2021 BeamOS Buster
```sh
sudo /usr/local/mrbeam_ledstrips/venv/bin/python3 -m pip install -e ~/sources/MrBeamLedStrips
sudo /usr/local/mrbeam_ledstrips/venv/bin/python -m pip install pytest==4.6.11 pytest-datafiles==2.0
cd ~/sources/MrBeamLedStrips
sudo /usr/local/mrbeam_ledstrips/venv/bin/python3 -m pytest -s
```

On the 2018 BeamOS Jessie
```sh
sudo python2 -m pip install -e ~/sources/MrBeamLedStrips
sudo python2 -m pip install pytest==4.6.11 pytest-datafiles==2.0
cd ~/sources/MrBeamLedStrips
sudo python2 -m pytest -s
```

amrsoll committed:
Fixes/Various

- Add opencv as a requirement (limited to py3)
- Fix dimming color when given an enum instead of int
- Fix rollback test as it didn't initialise state array correctly
- Fix undefined `fps` var

amrsoll committed:
Refactor/handle state -> handle command.

amrsoll committed:
Doc+Refactor/Various

- Use Constant OUTSIDE_LEDS instead of defining `involved_registers`
  every time.
- logger.warn() ==> logger.warning() (Deprecation warnings)
- use reversed(range(N)) instead of range(N-1, -1, -1)
- `for i in range(len(my_list))` ==> `for i, elm in enumerate(my_list)`
- `Colors.OFF if brightness==0 else dim_color(my_color, brightness)`
  ==>
  `dim_color(my_color, brightness)`
- Add notes and warnings for incoreect math used for animations
- Put some basic functions outside of class LEDs
- Make breathing() dependent on breathing_static() and not the other way
  around. NOTE: There is still a case where
  breathing_static(fade_in=True) uses breathing()
- copy-pasta ==> referencing other functions
- rewrite some formulas to be more readable
- Delete unused functions
- `set_fps(value)` ==> `self.fps = value` (uses property.setter)

amrsoll committed:
Set fps to minimum 1 Hz.

amrsoll committed:
Refactor/Various

- move repeated complex formula into function
- mute unused variable
- progress_pause() uses progress() instead of copy-pasta

amrsoll committed:
Fix breath factor.

amrsoll committed:
Refactor/range() -> enumerate() refactors.

amrsoll committed:
Refactor/Simplify png() logic

- Adds a new png file
- changed the breathin_orange to fit changes
- Remove more unused variables

amrsoll committed:
Refactor/Pytest cleanup.

ahmed-mrbeam commented:
@amrsoll Considering the change is huge, I would need the following to be done before merging:

- Create regression test plan for ledstrips using XRAY app here on Jira(covering all the ledstrips scenarios). Run all the tests and they pass before merging the PR back to develop . Please consult  @khaledsherkawi  for creating the tests 

- The test plan created to be reviewed and approved by  @Josef-MrBeam

amrsoll commented on mrbeam_ledstrips/state_animations.py:
```diff
@@ -716,405 +666,361 @@ def spread_spectrum(self, params):
+		elif command in COMMANDS['CUSTOM_COLOR']:
+			r, g, b = (int(args[i]) for i in range(3))
+			self.static_color(Color(r, g, b))
+			self.rollback_after_frames(self.frame, args.pop(0) if len(args)>0 else 0)
```
needs to be 
```python
if len(args) > 3:
    self.rollback_after_frames(self.frame, args[3])
```
the refactor removed the usage of `.pop()` for the first 3 values in the buffer

amrsoll commented:
I don't seem to be able to add @Josef-MrBeam as reviewer now, but I guess I would mention it here as a reminder :)
